verilog/analysis/checkers/disable_fork_rule_test.cc

```diff
@@ -0,0 +1,116 @@
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "verilog/analysis/checkers/disable_fork_rule.h"
+
+#include <initializer_list>
+
+#include "common/analysis/linter_test_utils.h"
+#include "common/analysis/syntax_tree_linter_test_utils.h"
+#include "common/text/symbol.h"
+#include "gtest/gtest.h"
+#include "verilog/CST/verilog_nonterminals.h"
+#include "verilog/CST/verilog_treebuilder_utils.h"
+#include "verilog/analysis/verilog_analyzer.h"
+#include "verilog/parser/verilog_token_enum.h"
+
+namespace verilog {
+namespace analysis {
+namespace {
+
+using verible::LintTestCase;
+using verible::RunLintTestCases;
+
+TEST(DisableStatementTest, FunctionPass) {
+  constexpr int kToken = TK_disable;
+  const std::initializer_list<LintTestCase> kDisableStatementTestCases = {
+      {""},
+      {"module m;\ninitial begin;\n", "fork\n", "begin\n#6;\nend\n",
+       "begin\n#3;\nend\n", "join_any\n", "disable fork;\n", "end\nendmodule"},
+      {"module m;\ninitial begin\n", "fork\n", "begin : foo\n",
+       "disable foo;\n", "end\n", "join_any\n", "end\nendmodule"},
+      {"module m;\ninitial begin\n", "fork\n", "begin : foo\n",
+       "begin : foo_2\n", "disable foo_2;\n", "end\n", "end\n", "join_any\n",
+       "end\nendmodule"},
+      {"module m;\ninitial begin\n", "fork\n", "begin : foo\n",
+       "begin : foo_2\n", "disable foo;\n", "end\n", "end\n", "join_any\n",
+       "end\nendmodule"},
+  };
+  RunLintTestCases<VerilogAnalyzer, DisableForkNoLabelsRule>(
+      kDisableStatementTestCases);
+}
+
+TEST(DisableStatementTest, FunctionFailures) {
+  constexpr int kToken = TK_disable;
+  const std::initializer_list<LintTestCase> kDisableStatementTestCases = {
+      {"module m;\ninitial begin\n",
+       "fork\n",
+       "begin\n#6;\nend\n",
+       "begin\n#3;\nend\n",
+       "join_any\n",
+       {kToken, "disable"},
+       " fork_invalid;\n",
+       "end\nendmodule"},
+      {"module m;\ninitial begin\n",
+       "fork:fork_label\n",
+       "begin\n#6;\nend\n",
+       "begin\n#3;\nend\n",
+       "join_any\n",
+       {kToken, "disable"},
+       " fork_label;\n",
+       "end\nendmodule"},
+      {"module m;\n",
+       "initial begin;\n",
+       "fork\n",
+       "begin : foo\n",
+       "end\n",
+       {kToken, "disable"},
+       " foo;\n",
+       "join_any\n",
+       "end\nendmodule"},
+      {"module m;\n",
+       "initial begin:foo;\n",
+       "fork\n",
+       "begin : fo\n",
+       {kToken, "disable"},
```

Comment:
I changed the tests, also I have added the test case you specified in the comment.
What I wanted to show by these tests is that disabling the non-sequential statements like `always_comb/final/initial` is illegal.

Reply:
Yes, I think we should separate out tests that deal with forks and other non-sequential statements so that we don't mix things in the test-cases and have too many variables tested at once (makes it easier to follow, and _if_ it fails at some point we know exactly what broke and don't have to try to find what aspect was failing).

I think we should have these in separate toplevel test-cases to make it clear to the reader what is being tested.
```c++
TEST(DisableStatementTest, ForkDisableStatements) { ... }
TEST(DisableStatementTest, NonSequentialDisableStatements) { ... }  // Maybe even split this in Initial, ...
```

Reply:
I decided that the two separate test cases are looking good just for now.
We are testing two main categories, and tearing them apart into smaller parts will lead to a situation where
there is only one test case for the test suite or we will repeat the tests